cdc/capture/http_handler.go

```diff
@@ -0,0 +1,544 @@
+// limitations under the License.
+
+package capture
+
+import (
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/ticdc/cdc/kv"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/owner"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/httputil"
+	"github.com/pingcap/ticdc/pkg/version"
+	"github.com/tikv/client-go/v2/oracle"
+	"go.etcd.io/etcd/clientv3/concurrency"
+)
+
+const (
+	// APIOpVarChangefeedState is the key of changefeed state in HTTP API
+	APIOpVarChangefeedState = "state"
+	// APIOpVarChangefeedID is the key of changefeed ID in HTTP API
+	APIOpVarChangefeedID = "changefeed_id"
+	// APIOpVarCaptureID is the key of capture ID in HTTP API
+	APIOpVarCaptureID = "capture_id"
+	// APIOpVarTableID is the key of table ID in HTTP API
+	APIOpVarTableID = "table_id"
+)
+
+type httpHandler struct {
+	capture *Capture
+}
+
+// NewHTTPHandler return a httpHandler for OpenAPI
+func NewHTTPHandler(capture *Capture) httpHandler {
+	return httpHandler{
+		capture: capture,
+	}
+}
+
+// ListChangefeed list all changgefeed in cdc cluster
+// @Summary List changefeed
+// @Description list all changefeed in cdc cluster
+// @Tags changefeed
+// @Accept  json
+// @Produce  json
+// @Param   state  query  string  false  "state"
+// @Success 200 {array} model.ChangefeedCommonInfo
+// @Failure 500 {object} model.HTTPError
+// @Router	/api/v1/changefeeds [get]
+func (h *httpHandler) ListChangefeed(c *gin.Context) {
+	state := c.Query(APIOpVarChangefeedState)
+	statuses, err := h.capture.etcdClient.GetAllChangeFeedStatus(c.Request.Context())
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	changefeedIDs := make(map[string]struct{}, len(statuses))
+	for cid := range statuses {
+		changefeedIDs[cid] = struct{}{}
+	}
+
+	resps := make([]*model.ChangefeedCommonInfo, 0)
+	for changefeedID := range changefeedIDs {
+		cfInfo, err := h.capture.etcdClient.GetChangeFeedInfo(c.Request.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+			return
+		}
+
+		if !httputil.IsFiltered(state, cfInfo.State) {
+			continue
+		}
+
+		cfStatus, _, err := h.capture.etcdClient.GetChangeFeedStatus(c.Request.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+			return
+		}
+
+		resp := &model.ChangefeedCommonInfo{
+			ID: changefeedID,
+		}
+
+		if cfInfo != nil {
+			resp.FeedState = cfInfo.State
+			resp.RunningError = cfInfo.Error
+		}
+
+		if cfStatus != nil {
+			resp.CheckpointTSO = cfStatus.CheckpointTs
+			tm := oracle.GetTimeFromTS(cfStatus.CheckpointTs)
+			resp.CheckpointTime = model.JSONTime(tm)
+		}
+
+		resps = append(resps, resp)
```

Comment:
What do you expect the result to be when the changefeed is being deleted and you get an `ErrChangeFeedNotExists` somewhere above? Would this result be confusing to the user? More generally, since we are not doing a snapshot read, would the result cause any problem?

Reply:
Good advice.
I will continue the loop if `ErrChangeFeedNotExists.Equal(err) `.